redis/src/push_manager.rs

```diff
@@ -0,0 +1,254 @@
+use crate::{ErrorKind, FromRedisValue, PushKind, RedisError, RedisResult, Value};
+use arc_swap::ArcSwap;
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::mpsc::{SyncSender, TrySendError};
+use std::sync::Arc;
+
+/// Holds information about received Push data
+#[derive(Debug, Clone)]
+pub struct PushInfo {
+    /// Push Kind
+    pub kind: PushKind,
+    /// Data from push message
+    pub data: Vec<Value>,
+
+    /// Connection address to distinguish connections
+    pub con_addr: Arc<String>,
+}
+
+/// `PushSender` holds multiple types of mpsc channels
+#[derive(Clone)]
+pub enum PushSender {
+    /// Tokio mpsc UnboundedSender
+    #[cfg(feature = "aio")]
+    Tokio(tokio::sync::mpsc::UnboundedSender<PushInfo>),
+    /// Standard mpsc SyncSender
+    Standard(SyncSender<PushInfo>),
+}
+#[derive(Clone)]
+struct ChannelSubscription {
+    channel_id: usize,
+    sender: PushSender,
+}
+
+#[derive(Default)]
+struct SubscriptionHolder {
+    subscriptions: Arc<ArcSwap<HashMap<String, Vec<ChannelSubscription>>>>,
+    last_channel_id: Arc<AtomicUsize>,
+}
+
+impl SubscriptionHolder {
+    fn subscribe_to_channel(&self, channel_name: String, sender: PushSender) -> usize {
+        let channel_id = self.last_channel_id.fetch_add(1, Ordering::SeqCst);
+        let cs = ChannelSubscription { channel_id, sender };
+        self.subscriptions.rcu(|x| {
+            let mut subscriptions = HashMap::clone(x);
+            if let Some(sbs) = subscriptions.get_mut(&channel_name) {
+                sbs.push(cs.clone());
+            } else {
+                subscriptions.insert(channel_name.clone(), vec![cs.clone()]);
+            }
+            subscriptions
+        });
+        channel_id
+    }
+    fn unsubscribe_from_channel(&self, channel_name: String, channel_id: Option<usize>) -> bool {
+        let mut unsubscribed_from_all = false;
+        self.subscriptions.rcu(|x| {
+            unsubscribed_from_all = false;
+            let mut subscriptions = HashMap::clone(x);
+            if let Some(sbs) = subscriptions.get_mut(&channel_name) {
+                if let Some(channel_id) = channel_id {
+                    sbs.retain(|cs| cs.channel_id != channel_id)
+                } else {
+                    sbs.clear();
+                }
+                if sbs.is_empty() {
+                    subscriptions.remove(&channel_name);
+                    unsubscribed_from_all = true;
+                }
+            }
+            subscriptions
+        });
+        unsubscribed_from_all
+    }
+    fn try_send(&self, val: &Value, con_addr: &Arc<String>) -> RedisResult<()> {
+        let channel_name: String;
```

Comment:
Can you share your thoughts about the rationale for demuxing channels in our code?

The way I see it, the current design gives reasonable UX - 
Users who want to receive messages from separate channels on separate receivers get that out of the box, and users who want to receive on the same receiver can just pass clones of that receiver's sender.
But we're paying for this with a lot of complexity on our side - we have a two-tiered indexing system, for starters.

We can easily use just a single sender in the push manager (passed on initialization?), and let the user add on it a rather simple demuxing logic. This will still provide reasonable UX, but will significantly reduce the complexity of this class and allow for easier maintenance.

Reply:
I think this is the best solution we can offer for Pub/Sub but I agree there is complexity. I would like to keep it this way, more tests will make here easier for maintenance.

Reply:
Why do you think this is the best solution?

Reply:
I understand what you say as "user can have only one receiver to all channels they subscribe to".
If that is correct with current solution user has ability to diverge receivers per subscription and that's powerful also that is what i'm used to.